README.md
# quran README

This is the README for your extension "quran". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

-----------------------------------------------------------------------------------------------------------

## Working with Markdown

**Note:** You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+CMD+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux) or `Cmd+Space` (macOS) to see a list of Markdown snippets

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "quran" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

static/main.js
// This script will be run within the webview itself
// It cannot access the main VS Code APIs directly.
(function () {
    const vscode = acquireVsCodeApi();

    const oldState = vscode.getState();

    const counter = document.getElementById('lines-of-code-counter');
    console.log(oldState);
    let currentCount = (oldState && oldState.count) || 0;
    counter.textContent = currentCount;

    // setInterval(() => {
    //     counter.textContent = currentCount++;

    //     // Update state
    //     vscode.setState({ count: currentCount });

    //     // Alert the extension when the cat introduces a bug
    //     if (Math.random() < Math.min(0.001 * currentCount, 0.05)) {
    //         // Send a message back to the extension
    //         vscode.postMessage({
    //             command: 'alert',
    //             text: '🐛  on line ' + currentCount
    //         });
    //     }
    // }, 1000);

    // Handle messages sent from the extension to the webview
    window.addEventListener('message', event => {
        const message = event.data; // The json data that the extension sent
        switch (message.command) {
            case 'counter':
                currentCount = message.value;
                counter.textContent = currentCount;
                break;
            case 'refactor':
                currentCount = Math.ceil(currentCount +5);
                counter.textContent = currentCount;
                break;
        }
    });
}());

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import HolyQuranPanel from './Panel';
import {TanzilEngine} from './TanzilEngine';
import { DepNodeProvider, Dependency } from './nodeDependencies';

console.log('your extension "quran" js is loaded.');

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "quran" is now active!');
	const engine:TanzilEngine = TanzilEngine.init(context.extensionPath);

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	context.subscriptions.push(
		vscode.commands.registerCommand('holyQuran.start', () => {						
			// Create and show a new webview
			engine.showPage();
			// HolyQuranPanel.createOrShow(context.extensionPath);
			// const hq = new HolyQuranPanel(context.extensionPath);
			// context.subscriptions.push(hq);
			// const panel = vscode.window.createWebviewPanel(
			// 	'holyQuran', // Identifies the type of the webview. Used internally
			// 	'Cat Coding2', // Title of the panel displayed to the user
			// 	vscode.ViewColumn.One, // Editor column to show the new webview panel in.
			// 	{ // Webview options. More on these later.
			// 		// Enable javascript in the webview
			// 		enableScripts: true,

			// 		// And restrict the webview to only loading content from our extension's `media` directory.
			// 		localResourceRoots: [vscode.Uri.file(path.join(extensionPath, 'media'))]
			// 	}
			// );

			// // And set its HTML content
			// panel.webview.html = getWebviewContent();
		})
	);
	
	// Samples of `window.registerTreeDataProvider`
	const nodeDependenciesProvider = new DepNodeProvider();
	vscode.window.registerTreeDataProvider('nodeDependencies', nodeDependenciesProvider);
	vscode.commands.registerCommand('nodeDependencies.open', (surah: string) => vscode.window.showInformationMessage(`Successfully called edit entry on ${surah}.`));
	// vscode.commands.registerCommand('nodeDependencies.open', (node: Dependency) => vscode.window.showInformationMessage(`Successfully called edit entry on ${node.label}.`));
	
	context.subscriptions.push(
		vscode.commands.registerCommand('holyQuran.doRefactor', () => {
			// if (HolyQuranPanel.currentPanel) {
			// 	HolyQuranPanel.currentPanel.doRefactor();
			// }
		})
	);
	let disposable = vscode.commands.registerCommand('extension.helloWorld', () => {
		// The code you place here will be executed every time your command is executed

		// Display a message box to the user
		vscode.window.showInformationMessage('Hello World!');
	});

	context.subscriptions.push(disposable);
}

// function getWebviewContent() {
// 	console.log('geWebContent.called');
// 	return `<!DOCTYPE html>
//   <html lang="en">
//   <head>
// 	  <meta charset="UTF-8">
// 	  <meta name="viewport" content="width=device-width, initial-scale=1.0">
// 	  <title>Cat Coding</title>
//   </head>
//   <body>
// 	  <img src="https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif" width="300" />
//   </body>
//   </html>`;
// }
// this method is called when your extension is deactivated
export function deactivate() {}


src/tanzil.net.ts
export const Notice = 'Copyright (C) 2008-2011 Hamid Zarrabi-Zadeh, Source: tanzil.net';
export const QuranData = {
    Sura: [[],     
        // ['start', 'ayas', 'order', 'rukus', 'name', 'tname', 'ename', 'type']
        [0, 7, 5, 1, 'الفاتحة', "Al-Fatiha", 'The Opening', 'Meccan'], 
        [7, 286, 87, 40, 'البقرة', "Al-Baqara", 'The Heifer', 'Medinan'], [293, 200, 89, 20, 'آل عمران', "Al-i-Imran", 'The Family of Imran', 'Medinan'], [493, 176, 92, 24, 'النساء', "An-Nisa", 'The Women', 'Medinan'], [669, 120, 112, 16, 'المائدة', "Al-Ma'ida", 'The Table', 'Medinan'], [789, 165, 55, 20, 'الأنعام', "Al-An'am", 'The Cattle', 'Meccan'], [954, 206, 39, 24, 'الأعراف', "Al-A'raf", 'The Heights', 'Meccan'], [1160, 75, 88, 10, 'الأنفال', "Al-Anfal", 'The Spoils of War', 'Medinan'], [1235, 129, 113, 16, 'التوبة', "At-Tawba", 'The Repentance', 'Medinan'], [1364, 109, 51, 11, 'يونس', "Yunus", 'Jonah', 'Meccan'], [1473, 123, 52, 10, 'هود', "Hud", 'Hud', 'Meccan'], [1596, 111, 53, 12, 'يوسف', "Yusuf", 'Joseph', 'Meccan'], [1707, 43, 96, 6, 'الرعد', "Ar-Ra'd", 'The Thunder', 'Medinan'], [1750, 52, 72, 7, 'ابراهيم', "Ibrahim", 'Abraham', 'Meccan'], [1802, 99, 54, 6, 'الحجر', "Al-Hijr", 'The Stoneland', 'Meccan'], [1901, 128, 70, 16, 'النحل', "An-Nahl", 'The Honey Bees', 'Meccan'], [2029, 111, 50, 12, 'الإسراء', "Al-Isra", 'The Night Journey', 'Meccan'], [2140, 110, 69, 12, 'الكهف', "Al-Kahf", 'The Cave', 'Meccan'], [2250, 98, 44, 6, 'مريم', "Maryam", 'Mary', 'Meccan'], [2348, 135, 45, 8, 'طه', "Ta-Ha", 'Ta-Ha', 'Meccan'], [2483, 112, 73, 7, 'الأنبياء', "Al-Anbiya", 'The Prophets', 'Meccan'], [2595, 78, 103, 10, 'الحج', "Al-Hajj", 'The Pilgrimage', 'Medinan'], [2673, 118, 74, 6, 'المؤمنون', "Al-Mu'minun", 'The Believers', 'Meccan'], [2791, 64, 102, 9, 'النور', "An-Nur", 'The Light', 'Medinan'], [2855, 77, 42, 6, 'الفرقان', "Al-Furqan", 'The Criterion', 'Meccan'], [2932, 227, 47, 11, 'الشعراء', "Ash-Shu'ara", 'The Poets', 'Meccan'], [3159, 93, 48, 7, 'النمل', "An-Naml", 'The Ant', 'Meccan'], [3252, 88, 49, 9, 'القصص', "Al-Qasas", 'The Stories', 'Meccan'], [3340, 69, 85, 7, 'العنكبوت', "Al-Ankabut", 'The Spider', 'Meccan'], [3409, 60, 84, 6, 'الروم', "Ar-Rum", 'The Romans', 'Meccan'], [3469, 34, 57, 4, 'لقمان', "Luqman", 'Luqman', 'Meccan'], [3503, 30, 75, 3, 'السجدة', "As-Sajda", 'The Prostration', 'Meccan'], [3533, 73, 90, 9, 'الأحزاب', "Al-Ahzab", 'The Clans', 'Medinan'], [3606, 54, 58, 6, 'سبإ', "Saba", 'Sheba', 'Meccan'], [3660, 45, 43, 5, 'فاطر', "Fatir", 'The Originator', 'Meccan'], [3705, 83, 41, 5, 'يس', "Ya-Sin", 'Yaseen', 'Meccan'], [3788, 182, 56, 5, 'الصافات', "As-Saffat", 'Drawn up in Ranks', 'Meccan'], [3970, 88, 38, 5, 'ص', "Sad", 'The Letter Sad', 'Meccan'], [4058, 75, 59, 8, 'الزمر', "Az-Zumar", 'The Troops', 'Meccan'], [4133, 85, 60, 9, 'غافر', "Ghafir", 'The Forgiver', 'Meccan'], [4218, 54, 61, 6, 'فصلت', "Fussilat", 'Explained in Detail', 'Meccan'], [4272, 53, 62, 5, 'الشورى', "Ash-Shura", 'The Consultation', 'Meccan'], [4325, 89, 63, 7, 'الزخرف', "Az-Zukhruf", 'The Ornaments of Gold', 'Meccan'], [4414, 59, 64, 3, 'الدخان', "Ad-Dukhan", 'The Smoke', 'Meccan'], [4473, 37, 65, 4, 'الجاثية', "Al-Jathiya", 'Crouching', 'Meccan'], [4510, 35, 66, 4, 'الأحقاف', "Al-Ahqaf", 'The Dunes', 'Meccan'], [4545, 38, 95, 4, 'محمد', "Muhammad", 'Muhammad', 'Medinan'], [4583, 29, 111, 4, 'الفتح', "Al-Fath", 'The Victory', 'Medinan'], [4612, 18, 106, 2, 'الحجرات', "Al-Hujurat", 'The Inner Apartments', 'Medinan'], [4630, 45, 34, 3, 'ق', "Qaf", 'The Letter Qaf', 'Meccan'], [4675, 60, 67, 3, 'الذاريات', "Adh-Dhariyat", 'The Winnowing Winds', 'Meccan'], [4735, 49, 76, 2, 'الطور', "At-Tur", 'The Mount', 'Meccan'], [4784, 62, 23, 3, 'النجم', "An-Najm", 'The Star', 'Meccan'], [4846, 55, 37, 3, 'القمر', "Al-Qamar", 'The Moon', 'Meccan'], [4901, 78, 97, 3, 'الرحمن', "Ar-Rahman", 'The Beneficent', 'Medinan'], [4979, 96, 46, 3, 'الواقعة', "Al-Waqi'a", 'The Inevitable', 'Meccan'], [5075, 29, 94, 4, 'الحديد', "Al-Hadid", 'The Iron', 'Medinan'], [5104, 22, 105, 3, 'المجادلة', "Al-Mujadila", 'The Pleading', 'Medinan'], [5126, 24, 101, 3, 'الحشر', "Al-Hashr", 'The Exile', 'Medinan'], [5150, 13, 91, 2, 'الممتحنة', "Al-Mumtahina", 'Examining Her', 'Medinan'], [5163, 14, 109, 2, 'الصف', "As-Saff", 'The Ranks', 'Medinan'], [5177, 11, 110, 2, 'الجمعة', "Al-Jumu'a", 'Friday', 'Medinan'], [5188, 11, 104, 2, 'المنافقون', "Al-Munafiqun", 'The Hypocrites', 'Medinan'], [5199, 18, 108, 2, 'التغابن', "At-Taghabun", 'Mutual Disillusion', 'Medinan'], [5217, 12, 99, 2, 'الطلاق', "At-Talaq", 'Divorce', 'Medinan'], [5229, 12, 107, 2, 'التحريم', "At-Tahrim", 'The Prohibition', 'Medinan'], [5241, 30, 77, 2, 'الملك', "Al-Mulk", 'The Sovereignty', 'Meccan'], [5271, 52, 2, 2, 'القلم', "Al-Qalam", 'The Pen', 'Meccan'], [5323, 52, 78, 2, 'الحاقة', "Al-Haqqa", 'The Reality', 'Meccan'], [5375, 44, 79, 2, 'المعارج', "Al-Ma'arij", 'The Ascending Stairways', 'Meccan'], [5419, 28, 71, 2, 'نوح', "Nuh", 'Noah', 'Meccan'], [5447, 28, 40, 2, 'الجن', "Al-Jinn", 'The Jinn', 'Meccan'], [5475, 20, 3, 2, 'المزمل', "Al-Muzzammil", 'The Enshrouded One', 'Meccan'], [5495, 56, 4, 2, 'المدثر', "Al-Muddathir", 'The Cloaked One', 'Meccan'], [5551, 40, 31, 2, 'القيامة', "Al-Qiyama", 'The Resurrection', 'Meccan'], [5591, 31, 98, 2, 'الانسان', "Al-Insan", 'Human', 'Medinan'], [5622, 50, 33, 2, 'المرسلات', "Al-Mursalat", 'The Emissaries', 'Meccan'], [5672, 40, 80, 2, 'النبإ', "An-Naba'", 'The Announcement', 'Meccan'], [5712, 46, 81, 2, 'النازعات', "An-Nazi'at", 'Those Who Drag Forth', 'Meccan'], [5758, 42, 24, 1, 'عبس', "Abasa", 'He Frowned', 'Meccan'], [5800, 29, 7, 1, 'التكوير', "At-Takwir", 'The Folding Up', 'Meccan'], [5829, 19, 82, 1, 'الإنفطار', "Al-Infitar", 'The Cleaving', 'Meccan'], [5848, 36, 86, 1, 'المطففين', "Al-Mutaffifin", 'Defrauding', 'Meccan'], [5884, 25, 83, 1, 'الإنشقاق', "Al-Inshiqaq", 'The Splitting Open', 'Meccan'], [5909, 22, 27, 1, 'البروج', "Al-Buruj", 'The Constellations', 'Meccan'], [5931, 17, 36, 1, 'الطارق', "At-Tariq", 'The Morning Star', 'Meccan'], [5948, 19, 8, 1, 'الأعلى', "Al-A'la", 'The Most High', 'Meccan'], [5967, 26, 68, 1, 'الغاشية', "Al-Ghashiya", 'The Overwhelming', 'Meccan'], [5993, 30, 10, 1, 'الفجر', "Al-Fajr", 'The Dawn', 'Meccan'], [6023, 20, 35, 1, 'البلد', "Al-Balad", 'The City', 'Meccan'], [6043, 15, 26, 1, 'الشمس', "Ash-Shams", 'The Sun', 'Meccan'], [6058, 21, 9, 1, 'الليل', "Al-Lail", 'The Night', 'Meccan'], [6079, 11, 11, 1, 'الضحى', "Ad-Dhuha", 'The Morning Hours', 'Meccan'], [6090, 8, 12, 1, 'الشرح', "Ash-Sharh", 'The Consolation', 'Meccan'], [6098, 8, 28, 1, 'التين', "At-Tin", 'The Fig', 'Meccan'], [6106, 19, 1, 1, 'العلق', "Al-Alaq", 'The Clot', 'Meccan'], [6125, 5, 25, 1, 'القدر', "Al-Qadr", 'The Power, Fate', 'Meccan'], [6130, 8, 100, 1, 'البينة', "Al-Bayyina", 'The Evidence', 'Medinan'], [6138, 8, 93, 1, 'الزلزلة', "Az-Zalzala", 'The Earthquake', 'Medinan'], [6146, 11, 14, 1, 'العاديات', "Al-Adiyat", 'The Chargers', 'Meccan'], [6157, 11, 30, 1, 'القارعة', "Al-Qari'a", 'The Calamity', 'Meccan'], [6168, 8, 16, 1, 'التكاثر', "At-Takathur", 'Competition', 'Meccan'], [6176, 3, 13, 1, 'العصر', "Al-Asr", 'The Time', 'Meccan'], [6179, 9, 32, 1, 'الهمزة', "Al-Humaza", 'The Traducer', 'Meccan'], [6188, 5, 19, 1, 'الفيل', "Al-Fil", 'The Elephant', 'Meccan'], [6193, 4, 29, 1, 'قريش', "Quraysh", 'Quraysh', 'Meccan'], [6197, 7, 17, 1, 'الماعون', "Al-Ma'un", 'Almsgiving', 'Meccan'], [6204, 3, 15, 1, 'الكوثر', "Al-Kawthar", 'Abundance', 'Meccan'], [6207, 6, 18, 1, 'الكافرون', "Al-Kafirun", 'The Disbelievers', 'Meccan'], [6213, 3, 114, 1, 'النصر', "An-Nasr", 'Divine Support', 'Medinan'], [6216, 5, 6, 1, 'المسد', "Al-Masad", 'The Palm Fibre', 'Meccan'], [6221, 4, 22, 1, 'الإخلاص', "Al-Ikhlas", 'Purity of Faith', 'Meccan'], [6225, 5, 20, 1, 'الفلق', "Al-Falaq", 'The Dawn', 'Meccan'], [6230, 6, 21, 1, 'الناس', "An-Nas", 'Mankind', 'Meccan'], [6236, 1]],

        Juz: [[], [1, 1], [2, 142], [2, 253], [3, 93], [4, 24], [4, 148], [5, 82], [6, 111], [7, 88], [8, 41], [9, 93], [11, 6], [12, 53], [15, 1], [17, 1], [18, 75], [21, 1], [23, 1], [25, 21], [27, 56], [29, 46], [33, 31], [36, 28], [39, 32], [41, 47], [46, 1], [51, 31], [58, 1], [67, 1], [78, 1], [115, 1]],
        Page: [[], [1, 1], [2, 1], [2, 6], [2, 17], [2, 25], [2, 30], [2, 38], [2, 49], [2, 58], [2, 62], [2, 70], [2, 77], [2, 84], [2, 89], [2, 94], [2, 102], [2, 106], [2, 113], [2, 120], [2, 127], [2, 135], [2, 142], [2, 146], [2, 154], [2, 164], [2, 170], [2, 177], [2, 182], [2, 187], [2, 191], [2, 197], [2, 203], [2, 211], [2, 216], [2, 220], [2, 225], [2, 231], [2, 234], [2, 238], [2, 246], [2, 249], [2, 253], [2, 257], [2, 260], [2, 265], [2, 270], [2, 275], [2, 282], [2, 283], [3, 1], [3, 10], [3, 16], [3, 23], [3, 30], [3, 38], [3, 46], [3, 53], [3, 62], [3, 71], [3, 78], [3, 84], [3, 92], [3, 101], [3, 109], [3, 116], [3, 122], [3, 133], [3, 141], [3, 149], [3, 154], [3, 158], [3, 166], [3, 174], [3, 181], [3, 187], [3, 195], [4, 1], [4, 7], [4, 12], [4, 15], [4, 20], [4, 24], [4, 27], [4, 34], [4, 38], [4, 45], [4, 52], [4, 60], [4, 66], [4, 75], [4, 80], [4, 87], [4, 92], [4, 95], [4, 102], [4, 106], [4, 114], [4, 122], [4, 128], [4, 135], [4, 141], [4, 148], [4, 155], [4, 163], [4, 171], [4, 176], [5, 3], [5, 6], [5, 10], [5, 14], [5, 18], [5, 24], [5, 32], [5, 37], [5, 42], [5, 46], [5, 51], [5, 58], [5, 65], [5, 71], [5, 77], [5, 83], [5, 90], [5, 96], [5, 104], [5, 109], [5, 114], [6, 1], [6, 9], [6, 19], [6, 28], [6, 36], [6, 45], [6, 53], [6, 60], [6, 69], [6, 74], [6, 82], [6, 91], [6, 95], [6, 102], [6, 111], [6, 119], [6, 125], [6, 132], [6, 138], [6, 143], [6, 147], [6, 152], [6, 158], [7, 1], [7, 12], [7, 23], [7, 31], [7, 38], [7, 44], [7, 52], [7, 58], [7, 68], [7, 74], [7, 82], [7, 88], [7, 96], [7, 105], [7, 121], [7, 131], [7, 138], [7, 144], [7, 150], [7, 156], [7, 160], [7, 164], [7, 171], [7, 179], [7, 188], [7, 196], [8, 1], [8, 9], [8, 17], [8, 26], [8, 34], [8, 41], [8, 46], [8, 53], [8, 62], [8, 70], [9, 1], [9, 7], [9, 14], [9, 21], [9, 27], [9, 32], [9, 37], [9, 41], [9, 48], [9, 55], [9, 62], [9, 69], [9, 73], [9, 80], [9, 87], [9, 94], [9, 100], [9, 107], [9, 112], [9, 118], [9, 123], [10, 1], [10, 7], [10, 15], [10, 21], [10, 26], [10, 34], [10, 43], [10, 54], [10, 62], [10, 71], [10, 79], [10, 89], [10, 98], [10, 107], [11, 6], [11, 13], [11, 20], [11, 29], [11, 38], [11, 46], [11, 54], [11, 63], [11, 72], [11, 82], [11, 89], [11, 98], [11, 109], [11, 118], [12, 5], [12, 15], [12, 23], [12, 31], [12, 38], [12, 44], [12, 53], [12, 64], [12, 70], [12, 79], [12, 87], [12, 96], [12, 104], [13, 1], [13, 6], [13, 14], [13, 19], [13, 29], [13, 35], [13, 43], [14, 6], [14, 11], [14, 19], [14, 25], [14, 34], [14, 43], [15, 1], [15, 16], [15, 32], [15, 52], [15, 71], [15, 91], [16, 7], [16, 15], [16, 27], [16, 35], [16, 43], [16, 55], [16, 65], [16, 73], [16, 80], [16, 88], [16, 94], [16, 103], [16, 111], [16, 119], [17, 1], [17, 8], [17, 18], [17, 28], [17, 39], [17, 50], [17, 59], [17, 67], [17, 76], [17, 87], [17, 97], [17, 105], [18, 5], [18, 16], [18, 21], [18, 28], [18, 35], [18, 46], [18, 54], [18, 62], [18, 75], [18, 84], [18, 98], [19, 1], [19, 12], [19, 26], [19, 39], [19, 52], [19, 65], [19, 77], [19, 96], [20, 13], [20, 38], [20, 52], [20, 65], [20, 77], [20, 88], [20, 99], [20, 114], [20, 126], [21, 1], [21, 11], [21, 25], [21, 36], [21, 45], [21, 58], [21, 73], [21, 82], [21, 91], [21, 102], [22, 1], [22, 6], [22, 16], [22, 24], [22, 31], [22, 39], [22, 47], [22, 56], [22, 65], [22, 73], [23, 1], [23, 18], [23, 28], [23, 43], [23, 60], [23, 75], [23, 90], [23, 105], [24, 1], [24, 11], [24, 21], [24, 28], [24, 32], [24, 37], [24, 44], [24, 54], [24, 59], [24, 62], [25, 3], [25, 12], [25, 21], [25, 33], [25, 44], [25, 56], [25, 68], [26, 1], [26, 20], [26, 40], [26, 61], [26, 84], [26, 112], [26, 137], [26, 160], [26, 184], [26, 207], [27, 1], [27, 14], [27, 23], [27, 36], [27, 45], [27, 56], [27, 64], [27, 77], [27, 89], [28, 6], [28, 14], [28, 22], [28, 29], [28, 36], [28, 44], [28, 51], [28, 60], [28, 71], [28, 78], [28, 85], [29, 7], [29, 15], [29, 24], [29, 31], [29, 39], [29, 46], [29, 53], [29, 64], [30, 6], [30, 16], [30, 25], [30, 33], [30, 42], [30, 51], [31, 1], [31, 12], [31, 20], [31, 29], [32, 1], [32, 12], [32, 21], [33, 1], [33, 7], [33, 16], [33, 23], [33, 31], [33, 36], [33, 44], [33, 51], [33, 55], [33, 63], [34, 1], [34, 8], [34, 15], [34, 23], [34, 32], [34, 40], [34, 49], [35, 4], [35, 12], [35, 19], [35, 31], [35, 39], [35, 45], [36, 13], [36, 28], [36, 41], [36, 55], [36, 71], [37, 1], [37, 25], [37, 52], [37, 77], [37, 103], [37, 127], [37, 154], [38, 1], [38, 17], [38, 27], [38, 43], [38, 62], [38, 84], [39, 6], [39, 11], [39, 22], [39, 32], [39, 41], [39, 48], [39, 57], [39, 68], [39, 75], [40, 8], [40, 17], [40, 26], [40, 34], [40, 41], [40, 50], [40, 59], [40, 67], [40, 78], [41, 1], [41, 12], [41, 21], [41, 30], [41, 39], [41, 47], [42, 1], [42, 11], [42, 16], [42, 23], [42, 32], [42, 45], [42, 52], [43, 11], [43, 23], [43, 34], [43, 48], [43, 61], [43, 74], [44, 1], [44, 19], [44, 40], [45, 1], [45, 14], [45, 23], [45, 33], [46, 6], [46, 15], [46, 21], [46, 29], [47, 1], [47, 12], [47, 20], [47, 30], [48, 1], [48, 10], [48, 16], [48, 24], [48, 29], [49, 5], [49, 12], [50, 1], [50, 16], [50, 36], [51, 7], [51, 31], [51, 52], [52, 15], [52, 32], [53, 1], [53, 27], [53, 45], [54, 7], [54, 28], [54, 50], [55, 17], [55, 41], [55, 68], [56, 17], [56, 51], [56, 77], [57, 4], [57, 12], [57, 19], [57, 25], [58, 1], [58, 7], [58, 12], [58, 22], [59, 4], [59, 10], [59, 17], [60, 1], [60, 6], [60, 12], [61, 6], [62, 1], [62, 9], [63, 5], [64, 1], [64, 10], [65, 1], [65, 6], [66, 1], [66, 8], [67, 1], [67, 13], [67, 27], [68, 16], [68, 43], [69, 9], [69, 35], [70, 11], [70, 40], [71, 11], [72, 1], [72, 14], [73, 1], [73, 20], [74, 18], [74, 48], [75, 20], [76, 6], [76, 26], [77, 20], [78, 1], [78, 31], [79, 16], [80, 1], [81, 1], [82, 1], [83, 7], [83, 35], [85, 1], [86, 1], [87, 16], [89, 1], [89, 24], [91, 1], [92, 15], [95, 1], [97, 1], [98, 8], [100, 10], [103, 1], [106, 1], [109, 1], [112, 1], [115, 1]]
    };


src/WebviewManager.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
// import throttle = require('lodash.throttle'); 
import HolyQuranPanel from './Panel';
import { DisposableComponent, dispose } from './DisposableComponent';



// function setShowRunningTasks_theLast(context:vscode.ExtensionContext) {
// 	setTimeout(() => {
// 		vscode.commands.getCommands().then((cmds)=>{
// 			console.log( cmds );
// 		});
// 	}, 2000);
// }

let myCounter : number = 0;

export class WebviewManager extends DisposableComponent {
	private myPanels : DisposableComponent;
	private readonly extensionPath: string;
	// private myPanels : StatusbarTask[] = [];
	constructor(extensionPath: string) {
		super();
		this.extensionPath = extensionPath;
		this.myPanels = new DisposableComponent();
		this.addDisposable(this.myPanels);
		// this.addDisposable(vscode.workspace.onDidChangeConfiguration(() => {
		// 	this.reload();
		// }));
		// this.addConfigureTaskButton();
		
		// this.reload();
	}

	AddPage(){
		const hq = new HolyQuranPanel(this.extensionPath);
		this.myPanels.addDisposable(hq);
	}
}

src/Panel.ts
import * as path from 'path';
import * as vscode from 'vscode';

const cats = {
	'Coding Cat': 'https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif',
	'Compiling Cat': 'https://media.giphy.com/media/mlvseq9yvZhba/giphy.gif',
	'Testing Cat': 'https://media.giphy.com/media/3oriO0OEd9QIDdllqo/giphy.gif'
};

/**
 * Manages cat coding webview panels
 */
export default class HolyQuranPanel {
	/**
	 * Track the currently panel. Only allow a single panel to exist at a time.
	 */
	// public static currentPanel: HolyQuranPanel | undefined;

	public static readonly viewType = 'holyQuran';

	private readonly _webview: vscode.WebviewPanel;
	private readonly _extensionPath: string;
	private _disposables: vscode.Disposable[] = [];

	get ready() {return this._webview.visible;}

	public static createOrShow(extensionPath: string) {
		const column = vscode.window.activeTextEditor
			? vscode.window.activeTextEditor.viewColumn
			: undefined;

		// If we already have a panel, show it.
		// if (HolyQuranPanel.currentPanel) {
		// 	HolyQuranPanel.currentPanel._panel.reveal(column);
		// 	return;
		// }

		// Otherwise, create a new panel.
		const panel = vscode.window.createWebviewPanel(
			HolyQuranPanel.viewType,
			'Cat CodingT',
			column || vscode.ViewColumn.One,
			{
				// Enable javascript in the webview
                enableScripts: true,
                enableFindWidget: true,
                retainContextWhenHidden: true,

				// And restrict the webview to only loading content from our extension's `media` directory.
				localResourceRoots: [
                    vscode.Uri.file(path.join(extensionPath, 'static')),
                    vscode.Uri.file(path.join(extensionPath, 'out', 'public'))
                ]
			}
		);

		// HolyQuranPanel.currentPanel = new HolyQuranPanel(panel, extensionPath);
	}

	// public static revive(panel: vscode.WebviewPanel, extensionPath: string) {
	// 	HolyQuranPanel.currentPanel = new HolyQuranPanel(panel, extensionPath);
	// }

	public constructor(extensionPath: string) {
		const column = vscode.window.activeTextEditor
			? vscode.window.activeTextEditor.viewColumn
			: undefined;

		this._webview = vscode.window.createWebviewPanel(
			HolyQuranPanel.viewType,
			'Cat CodingT',
			column || vscode.ViewColumn.One,
			{
				// Enable javascript in the webview
                enableScripts: true,
                enableFindWidget: true,
                retainContextWhenHidden: true,

				// And restrict the webview to only loading content from our extension's `media` directory.
				localResourceRoots: [
                    vscode.Uri.file(path.join(extensionPath, 'static')),
                    vscode.Uri.file(path.join(extensionPath, 'out', 'public'))
                ]
			}
		);
		this._extensionPath = extensionPath;

		// Set the webview's initial html content
		this._update();

		// Listen for when the panel is disposed
		// This happens when the user closes the panel or when the panel is closed programatically
		this._webview.onDidDispose(() => this.dispose(), null, this._disposables);

		// Update the content based on view changes
		this._webview.onDidChangeViewState(
			e => {
				if (this._webview.visible) {
					this._update();
				}
			},
			null,
			this._disposables
		);

		// Handle messages from the webview
		this._webview.webview.onDidReceiveMessage(
			message => {
				switch (message.command) {
					case 'alert':
						vscode.window.showErrorMessage(message.text);
						return;
				}
			},
			null,
			this._disposables
		);
	}

	public postMessage(message: any) {
		// Send a message to the webview webview.
		// You can send any JSON serializable data.
		if(this.ready) {
			this._webview.webview.postMessage(message /* { command: 'refactor' } */);
		}
	}

	public dispose() {
		// HolyQuranPanel.currentPanel = undefined;

		// Clean up our resources
		this._webview.dispose();

		while (this._disposables.length) {
			const x = this._disposables.pop();
			if (x) {
				x.dispose();
			}
		}
	}

	private _update() {
		const webview = this._webview.webview;

		// Vary the webview's content based on where it is located in the editor.
		switch (this._webview.viewColumn) {
			case vscode.ViewColumn.Two:
				this._updateForCat(webview, 'Compiling Cat');
				return;

			case vscode.ViewColumn.Three:
				this._updateForCat(webview, 'Testing Cat');
				return;

			case vscode.ViewColumn.One:
			default:
				this._updateForCat(webview, 'Coding Cat');
				return;
		}
	}

	private _updateForCat(webview: vscode.Webview, catName: keyof typeof cats) {
		this._webview.title = catName;
		this._webview.webview.html = this._getHtmlForWebview(webview, cats[catName]);
	}

	private _getHtmlForWebview(webview: vscode.Webview, catGifPath: string) {
		// Local path to main script run in the webview
		const scriptPathOnDisk = vscode.Uri.file(
			path.join(this._extensionPath, 'static', 'main.js')
		);

		// And the uri we use to load this script in the webview
		const scriptUri = webview.asWebviewUri(scriptPathOnDisk);

		// Use a nonce to whitelist which scripts can be run
		const nonce = getNonce();

		return `<!DOCTYPE html>
            <html lang="en">
            <head>
                <meta charset="UTF-8">

                <!--
                Use a content security policy to only allow loading images from https or from our extension directory,
                and only allow scripts that have a specific nonce.
                -->
                <meta http-equiv="Content-Security-Policy" content="default-src 'none'; img-src ${webview.cspSource} https:; script-src 'nonce-${nonce}';">

                <meta name="viewport" content="width=device-width, initial-scale=1.0">
                <title>Cat Coding2</title>
            </head>
            <body>
                <!-- img src="${catGifPath}" width="300" / -->
                <h1 id="lines-of-code-counter">0</h1>

                <script nonce="${nonce}" src="${scriptUri}"></script>
            </body>
            </html>`;
	}
}

function getNonce() {
	let text = '';
	const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	for (let i = 0; i < 32; i++) {
		text += possible.charAt(Math.floor(Math.random() * possible.length));
	}
	return text;
}
